pkg/proxy/handlers/spacelister.go

```diff
@@ -0,0 +1,157 @@
+		if len(workspaces) != 1 {
+			// not found
+			r := schema.GroupResource{Group: "toolchain.dev.openshift.com", Resource: "workspaces"}
+			return errorResponse(ctx, apierrors.NewNotFound(r, workspaceName), http.StatusNotFound)
+		}
+		return getWorkspaceResponse(ctx, workspaces[0])
+	}
+
+	return listWorkspaceResponse(ctx, workspaces)
+}
+
+func errorResponse(ctx echo.Context, err *apierrors.StatusError, code int) error {
+	ctx.Logger().Error(err, err.Error())
+	ctx.Response().Writer.Header().Set("Content-Type", "application/json")
+	ctx.Response().Writer.WriteHeader(code)
+	return json.NewEncoder(ctx.Response().Writer).Encode(err.ErrStatus)
+}
+
+func getWorkspaceResponse(ctx echo.Context, workspace toolchainv1alpha1.Workspace) error {
+	ctx.Response().Writer.Header().Set("Content-Type", "application/json")
+	ctx.Response().Writer.WriteHeader(http.StatusOK)
+	return json.NewEncoder(ctx.Response().Writer).Encode(workspace)
+}
+
+func listWorkspaceResponse(ctx echo.Context, workspaces []toolchainv1alpha1.Workspace) error {
+	workspaceList := &toolchainv1alpha1.WorkspaceList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "WorkspaceList",
+			APIVersion: "toolchain.dev.openshift.com/v1alpha1",
+		},
+		Items: workspaces,
+	}
+
+	ctx.Response().Writer.Header().Set("Content-Type", "application/json")
+	ctx.Response().Writer.WriteHeader(http.StatusOK)
+	return json.NewEncoder(ctx.Response().Writer).Encode(workspaceList)
+}
+
+func (s *SpaceLister) workspacesFromSpaceBindings(spaceBindings []*toolchainv1alpha1.SpaceBinding) ([]toolchainv1alpha1.Workspace, error) {
+	workspaces := []toolchainv1alpha1.Workspace{}
+	for _, spaceBinding := range spaceBindings {
+
+		if spaceBinding.Labels == nil {
+			continue
+		}
+
+		var ownerName string
+		spaceName := spaceBinding.Labels[toolchainv1alpha1.SpaceBindingSpaceLabelKey]
+		space, err := s.GetInformerServiceFunc().GetSpace(spaceName)
+		if err != nil {
+			// log error and continue so that the api behaves in a best effort manner
+			// ie. if a space isn't listed something went wrong but we still want to return the other spaces if possible
+			log.Errorf(nil, err, "unable to get space", "space", spaceName)
+			continue
+		}
+		if space.Labels != nil { // space may not be initialized yet
+			// TODO right now we get SpaceCreatorLabelKey but should get owner from Space once it's implemented
+			ownerName = space.Labels[toolchainv1alpha1.SpaceCreatorLabelKey]
+			if spaceName == "" { // space may not be initialized
+				continue
+			}
+		}
+
+		// TODO get namespaces from Space status once it's implemented
+		namespaces := []toolchainv1alpha1.SpaceNamespace{
+			{
+				Name: spaceName + "-tenant",
+				Type: "default",
+			},
+		}
+
+		workspace := commonproxy.NewWorkspace(spaceName,
+			commonproxy.WithNamespaces(namespaces),
+			commonproxy.WithOwner(ownerName),
+			commonproxy.WithRole(spaceBinding.Spec.SpaceRole),
+		)
+		workspaces = append(workspaces, *workspace)
+	}
+	return workspaces, nil
```

Comment:
Just curious - should this function list also sub-spaces? 
since those will not have their own spacebindings , but they inherit them from the parentSpace.

In case sub-space retrieval is needed it can be done using following label:
`toolchainv1alpha1.ParentSpaceLabelKey: spaceName`

Reply:
It might eventually but it was discussed in the design doc linked in https://issues.redhat.com/browse/CRT-1641 and the decision was to not include sub-spaces (at least at this point in time).